elephant/spectral.py

```diff
@@ -523,6 +529,543 @@ def multitaper_psd(signal, n_segments=1, len_segment=None,
+    data = np.asarray(signals)
+    if isinstance(signals, neo.AnalogSignal):
+        data = np.moveaxis(data, 0, 1)
+
+    # Number of data points in time series
+    length_signal = np.shape(data)[1]
+
+    # If the data is given as AnalogSignal, use its attribute to specify the
+    # sampling frequency
+    if hasattr(signals, 'sampling_rate'):
+        fs = signals.sampling_rate.rescale('Hz')
+
+    # If fs and peak resolution is pq.Quantity, get magnitude
+    if isinstance(fs, pq.quantity.Quantity):
+        fs = fs.rescale('Hz').magnitude
+
+    # Determine time-halfbandwidth product from given parameters
+    if peak_resolution is not None:
+        if isinstance(peak_resolution, pq.quantity.Quantity):
+            peak_resolution = peak_resolution.rescale('Hz').magnitude
+        if peak_resolution <= 0:
+            raise ValueError("peak_resolution must be positive")
+        nw = length_signal / fs * peak_resolution / 2
+        num_tapers = int(np.floor(2*nw) - 1)
+
+    if num_tapers is None:
+        num_tapers = int(np.floor(2*nw) - 1)
+    else:
+        if not isinstance(num_tapers, int):
+            raise TypeError("num_tapers must be integer")
+        if num_tapers <= 0:
+            raise ValueError("num_tapers must be positive")
+
+    # Get slepian functions (sym='False' used for spectral analysis)
+    slepian_fcts = scipy.signal.windows.dpss(M=length_signal,
+                                             NW=nw,
+                                             Kmax=num_tapers,
+                                             sym=False)
+
+    # Use broadcasting to match dime for point-wise multiplication
+    tapered_signals = (data[:, np.newaxis] * slepian_fcts)
+
+    if return_onesided:
+        # Determine frequencies for real Fourier transform
+        freqs = np.fft.rfftfreq(length_signal, 1 / fs)
+        # Determine real Fourier transform of tapered signal
+        spectrum_estimates = np.fft.rfft(tapered_signals, axis=-1)
+
+    else:
+        # Determine frequencies for full Fourier transform
+        freqs = np.fft.fftfreq(length_signal, 1 / fs)
+        # Determine full Fourier transform of tapered signal
+        spectrum_estimates = np.fft.fft(tapered_signals, axis=-1)
+
+    temp = (spectrum_estimates[np.newaxis, :, :, :]
+            * np.conjugate(spectrum_estimates[:, np.newaxis, :, :]))
+
+    # Average Fourier transform windowed signal
+    cross_spec = np.mean(temp, axis=-2, dtype=np.complex64) / fs
+
+    if attach_units and isinstance(signals, pq.quantity.Quantity):
+        freqs = freqs * pq.Hz
+        cross_spec = cross_spec * signals.units * signals.units / pq.Hz
+
+    return freqs, cross_spec
+
+
+def _segmented_apply_func(data, func,  n_segments=1, len_segment=None,
+                          frequency_resolution=None, overlap=0.5,
+                          func_params_dict=None):
+    """
+    Estimate a spectral measure of a signal by applying it to segments of a
```

Comment:
This docstring should be removed and relevant parts moved to the respective public function. What we have been using is adding a short comment block at the beginning of the function with a short description of what the private function is doing (e.g., this function cuts `data` into several overlapping segments, applies `func` to each segment and average the obtained estimates across segments. Parameters of `func` are passed through the dictionary `func_params_dict`). 

This is complemented by a short description of the role of relevant parameters (e.g., the length of the segments are determined by parameters `n_segments`, `len_segment` or `frequency_resolution`).

The objective is to view a short overview of what the code is doing.

Reply:
We would like to keep it as is